pkg/crypto/certmanager/storage/awssecretsmanager/impl.go

```diff
@@ -0,0 +1,224 @@
+	secretId string) (map[string]string, error) {
+	input := secretsmanager.GetSecretValueInput{SecretId: aws.String(secretId)}
+	output, err := awsService.GetSecretValue(&input)
+	if err != nil {
+		return nil,
+			fmt.Errorf("error calling secretsmanager:GetSecretValue: %s", err)
+	}
+	if output.SecretString == nil {
+		return nil, errors.New("no SecretString in secret")
+	}
+	secret := []byte(*output.SecretString)
+	var secrets map[string]string
+	if err := json.Unmarshal(secret, &secrets); err != nil {
+		return nil, fmt.Errorf("error unmarshaling secret: %s", err)
+	}
+	return secrets, nil
+}
+
+func newLS(secretId string, logger log.DebugLogger) (*LockingStorer, error) {
+	awsService, err := getAwsService(secretId)
+	if err != nil {
+		return nil, err
+	}
+	return &LockingStorer{
+		awsService: awsService,
+		logger:     logger,
+		secretId:   secretId,
+	}, nil
+}
+
+func putAwsSecret(awsService *secretsmanager.SecretsManager,
+	secretId string, secrets map[string]string) error {
+	secretString, err := json.Marshal(secrets)
+	if err != nil {
+		return err
+	}
+	input := secretsmanager.PutSecretValueInput{
+		SecretId:     aws.String(secretId),
+		SecretString: aws.String(string(secretString)),
+	}
+	output, err := awsService.PutSecretValue(&input)
+	if err != nil {
+		return fmt.Errorf("error calling secretsmanager:PutSecretValue: %s",
+			err)
+	}
+	for _, versionStage := range output.VersionStages {
+		if *versionStage == "AWSCURRENT" {
+			return nil
+		}
+	}
+	return errors.New("no AWSCURRENT version stage associated")
+}
+
+func (ls *LockingStorer) read() (*certmanager.Certificate, error) {
+	keyMap, err := getAwsSecret(ls.awsService, ls.secretId)
+	if err != nil {
+		return nil, err
+	}
+	certPEM := &bytes.Buffer{}
+	for index := 0; ; index++ {
+		certificateBase64 := keyMap[fmt.Sprintf("Certificate%d", index)]
+		if certificateBase64 == "" {
+			if index == 0 {
+				return nil, errors.New("no Certificate in map")
+			}
+			break // We've reached the end of the certificate chain.
+		}
+		certDER, err := base64.StdEncoding.DecodeString(
+			strings.Replace(certificateBase64, " ", "", -1))
+		if err != nil {
+			return nil, err
+		}
+		if index != 0 {
+			fmt.Fprintln(certPEM)
+		}
+		err = pem.Encode(certPEM, &pem.Block{
+			Type:  "CERTIFICATE",
+			Bytes: certDER,
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+	keyType := keyMap["KeyType"]
+	if keyType == "" {
+		return nil, errors.New("no KeyType in map")
+	}
+	privateKeyBase64 := keyMap["PrivateKey"]
+	if privateKeyBase64 == "" {
+		return nil, errors.New("no PrivateKey in map")
+	}
+	privateKey, err := base64.StdEncoding.DecodeString(
+		strings.Replace(privateKeyBase64, " ", "", -1))
+	if err != nil {
+		return nil, err
+	}
+	keyPEM := pem.EncodeToMemory(&pem.Block{
+		Type:  keyType + " PRIVATE KEY",
+		Bytes: privateKey,
+	})
+	ls.logger.Printf(
+		"read certificate from AWS Secrets Manager, SecretId: %s\n",
+		ls.secretId)
+	return &certmanager.Certificate{
```

Comment:
Before you return you should check that the key is actually a a key and that it's public key actually matches the public key in the certificate

Reply:
So, I don't see it as the responsibility of this layer to ensure data validity. The certificate manager will parse the data read from any `Storer`: https://github.com/Cloud-Foundations/golib/blob/10f118677bb6b8590f94203aa13602d8517aa0cd/pkg/crypto/certmanager/manager.go#L124